### DISCUSSION TITLE
Build some crates with nightly compiler and some with stable -- automatically

### DISCUSSION THREAD

**Discussion by cbrune** [bazelbuild/rules_rust#1955] [Q&A] [answered]
Hi,

First off -- thanks all for working on rules_rust.  It is really helpful.

My question is how to support building some crates with nightly and some with stable rustc without using any bazel command line arguments.  I know I can add `--@rules_rust//rust/toolchain/channel=nightly` to the command line to build a crate a with nightly.

For an example, say I have a WORKSPACE with two crates, `crate-stable` and and `crate-nightly`.  `crate-stable` is intended to be built with the stable compiler and `crate-nightly` is intended to built with the nightly compiler.

I would like to build (and my users to build) these crates without knowing about specifying any command line args.  So the following is the goal:

```
$ bazel build //demo:crate-stable     <--- uses stable compiler
$ bazel build //demo:crate-nightly    <--- uses nightly compiler
```

Question 1:  Has anyone does this already?
Question 2:  How to do it?

I gather it might involve writing a transition rule that transitions to use the nightly compiler.  Any examples of transitions and rules_rust to look at?

Cheers,
Curt

**Comment by UebelAndre** [comment, marked as answer]
You'd need to write a rule that transitions on `@rules_rust//rust/toolchain/channel`. We do transitions for the unit testing of this build setting [here](https://github.com/bazelbuild/rules_rust/blob/0.21.1/test/unit/channel_transitions/channel_transiitons_test.bzl) but to get into the details of actually writing a rule that transitions I'd see:
- https://bazel.build/extending/config#defining-1-2-transitions
- https://bazel.build/rules/lib/builtins/transition

This will allow you to have top level targets that are explicitly built for stable or nightly compilers.

**Reply by cbrune** [reply]
Thanks!  That help a lot.  Sorted it out.  Thanks again.
